servercfg/serverconf.go

```diff
@@ -479,17 +479,36 @@ func GetPublicIP() (string, error) {
 		iplist = append([]string{publicIpService}, iplist...)
 	}
 
+	// regular timeout
+	timeoutTarget := time.Duration(10) * time.Second
+
+	// our unit test sets a bad ip service to test the timeout and fallback mechanics
+	// if that env variable is set, lower the timeout so the test doesn't take forever
+	testCheck := os.Getenv("NETMAKER_TEST_BAD_IP_SERVICE")
+	if testCheck != "" {
+		// we also don't really need to be checking all the services in the list after the planned timeout
+		// we just need to make sure that it falls back to the following service, which will be a mock
+		iplist = iplist[:1]
+		iplist = append(iplist, os.Getenv("NETMAKER_TEST_IP_SERVICE"))
+	}
+
 	for _, ipserver := range iplist {
+		if testCheck != "" && strings.EqualFold(testCheck, ipserver) {
+			timeoutTarget = time.Duration(2) * time.Second
+		}
+
 		client := &http.Client{
-			Timeout: time.Second * 10,
+			Timeout: timeoutTarget,
 		}
-		resp, err := client.Get(ipserver)
+		var resp *http.Response
+		resp, err = client.Get(ipserver)
 		if err != nil {
 			continue
 		}
 		defer resp.Body.Close()
 		if resp.StatusCode == http.StatusOK {
-			bodyBytes, err := io.ReadAll(resp.Body)
+			var bodyBytes []byte
```

Comment:
no need to declare a var here  `bodyBytes, err :=io.ReadAll(resp.Body)`

Reply:
Well... this actually starts to touch on the bug that I'm fixing in the first place. The problem here is that we were checking for a non nil err afterwards that was... always nil... what I'm doing here is making sure that we are propagating that error value throughout the process so that by the time we check for a non-nil error, it will actually potentially contain a value 😅

Reply:
the non-nil check after completing the loop is not required.  It does not provide any additional information.  Once the loop is completed, if the endpoint string is still empty, the error "public address not found" should be returned.

Reply:
okay, so wouldn't we like to know.. why? as my unit test shows, with this PR, we do. I'd like to know the reason that this function failed if it failed. are you saying we're not handling the error from this function? if so then that'd be what I'll attack next.

Reply:
for the most part, this error is ignored by calling functions.  In the one case where it is checked, an err will  result in a fatalLog.

if we need the actual error for troubleshooting then just add a slog.Debug()  when the error occurs.